src/rgw/rgw_common.cc

```diff
@@ -1467,3 +1478,78 @@ int rgw_parse_op_type_list(const string& str, uint32_t *perm)
   return parse_list_of_flags(op_type_mapping, str, perm);
 }
 
+static int match_internal(const char* pattern, const char* input)
+{
+  while(true) {
+    if (*pattern == '\0' && *input == '\0')
+      return 1;
+    if (*pattern == '\0' || *input == '\0')
+      return 0;
+    if (*pattern == '*' && *(pattern + 1) == '\0' && *input != '\0')
+      return 1;
+    if (*pattern == '*' && *(pattern + 1) == '\0' && *input == '\0')
+      return 0;
+    if (*pattern == *input || *pattern == '?') {
+      pattern++;
+      input++;
+      continue;
+    }
+    if (*pattern == '*') {
+      if (*(pattern + 1) == *input)
+        pattern++;
+      else
+        input++;
+      continue;
+    }
+    return 0;
+  }
+  return 0;
+}
+
+static void tokenize(string& text, vector<string>& result)
+{
+  char_separator<char> sep(":");
+  tokenizer<char_separator<char> > tokens(text, sep);
+  for ( auto it : tokens)
+  {
+    result.push_back(it);
+  }
+}
+
+int match(string& pattern, string& input, int flag)
+{
+  if (flag & POLICY_ACTION || flag & POLICY_ARN) {
+    std::transform(pattern.begin(),
+                    pattern.end(),
+                    pattern.begin(),
+                    ::tolower);
+    std::transform(input.begin(),
+                   input.end(),
+                   input.begin(),
+                   ::tolower);
+  }
```

Comment:
I was thinking that rather than modify the string, one could have match_internal take a function that compares characters. Something like

template<typename F>

`static int match_internal(string_ref pattern, string_ref input, F&& eq)`

And then use it like `if (eq(a, b))` instead of `if (a == b)`

Then when calling match_internal you can either pass a function that does a a case-folding character comparison or a function that does a case sensitive compare.